### ISSUE TITLE
About the Meta Neural Network training method

### ISSUE THREAD

**Issue by auroua** [naszilla/bananas#5]
The code chip in method `banans` in file `nas_algorithms.py`
```
    while query <= total_queries:
        candidates = search_space.get_candidates(data,
                                                 acq_opt_type=acq_opt_type,
                                                 encode_paths=encode_paths,
                                                 allow_isomorphisms=allow_isomorphisms,
                                                 deterministic_loss=deterministic)
        xcandidates = np.array([c[1] for c in candidates])
        predictions = []
        # train an ensemble of neural networks
        train_error = 0
        for _ in range(num_ensemble):
            meta_neuralnet = MetaNeuralnet()
            train_error += meta_neuralnet.fit(xtrain, ytrain, **metann_params)
            # predict the validation loss of the candidate architectures
            predictions.append(np.squeeze(meta_neuralnet.predict(xcandidates)))
        train_error /= num_ensemble
        if verbose:
            print('Query {}, Meta neural net train error: {}'.format(query, train_error))
```

In each while loop, you created five totally new neural network and use the same training data to train the five new neural network. 
I can not understand why the five meta neural network is trained in this way. Using the same training data and training from scratch in each iteration.

**Comment by crwhite14** [comment]
Yes, this is true. Training five times (with a different random weight initialization and a different random shuffle of the training data) will give five predictions for each architecture, so we get a mean and uncertainty estimate for each architecture. This is a technique called ensembles:

Lakshminarayanan et al., Simple and Scalable Predictive Uncertainty Estimation using Deep Ensembles, 2017.
Choi et al, Ensemble of Deep Convolutional Neural Networks for Prognosis of Ischemic Stroke, 2017.
Beluch et al, The power of ensembles for active learning in image classification, 2018.

**Comment by auroua** [comment]
Thanks for you replay.
The code chip in method banans in file nas_algorithms.py
```
    data = search_space.generate_random_dataset(num=num_init, 
                                                encode_paths=encode_paths, 
                                                allow_isomorphisms=allow_isomorphisms,
                                                deterministic_loss=deterministic)
    specs = [d[0] for d in data]
    xtrain = np.array([d[1] for d in data])
    ytrain = np.array([d[2] for d in data])
    query = num_init + k

    while query <= total_queries:
        candidates = search_space.get_candidates(data,
                                                 acq_opt_type=acq_opt_type,
                                                 encode_paths=encode_paths,
                                                 allow_isomorphisms=allow_isomorphisms,
                                                 deterministic_loss=deterministic)
        xcandidates = np.array([c[1] for c in candidates])
        predictions = []
        # train an ensemble of neural networks
        train_error = 0
        for _ in range(num_ensemble):
            meta_neuralnet = MetaNeuralnet()
            train_error += meta_neuralnet.fit(xtrain,  ytrain, **metann_params)
            # predict the validation loss of the candidate architectures
            predictions.append(np.squeeze(meta_neuralnet.predict(xcandidates)))
        train_error /= num_ensemble
```
The `xtrain` object in your code sames never change. The `Algorithm 1 BANANAS` in your paper the step 1 same means the new sampled data in step 4 is added to the train data set, and use this new data set to train the ensemble of meta neural networks.

**Comment by crwhite14** [comment]
Sorry, now I understand your comment. Yes, we add the new data in each iteration of the while loop. I just pushed the changes. This was a typo that was introduced during our code cleanup, after we finished the experiments.

**Comment by auroua** [comment]
Thanks.
